src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,289 @@
+    s3selectEngine::json_object* json_object = nullptr;
+    const uint8_t* headers_buf;
+    uint32_t headers_len;
+};
+
+Napi::Value
+S3SelectNapi::Write(const Napi::CallbackInfo& info)
+{
+    Napi::Buffer<char> buffer = info[0].As<Napi::Buffer<char>>();
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        buffer.Data(),
+        buffer.Length(),
+        false,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::Value
+S3SelectNapi::Flush(const Napi::CallbackInfo& info)
+{
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        nullptr, /*No buffer for flush*/
+        0,
+        true,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::FunctionReference S3SelectNapi::constructor;
+
+Napi::Object
+S3SelectNapi::Init(Napi::Env env, Napi::Object exports)
+{
+    Napi::HandleScope scope(env);
+
+    Napi::Function func = DefineClass(
+        env,
+        "S3SelectNapi",
+        {
+            InstanceMethod("write", &S3SelectNapi::Write),
+            InstanceMethod("flush", &S3SelectNapi::Flush)
+        }
+    ); //end of DefineClass
+
+    constructor = Napi::Persistent(func);
+    constructor.SuppressDestruct();
+
+    exports.Set("S3Select", func);
+    return exports;
+}
+
+void
+s3select_napi(Napi::Env env, Napi::Object exports)
+{
+    S3SelectNapi::Init(env, exports);
+}
+
+std::string
+GetStringWithDefault(Napi::Object obj, std::string name, std::string dfault)
+{
+    if (obj.Has(name)) {
+        return obj.Get(name).ToString().Utf8Value();
+    }
+    return dfault;
+}
+
+S3SelectNapi::S3SelectNapi(const Napi::CallbackInfo& info)
+    : Napi::ObjectWrap<S3SelectNapi>(info)
+{
+    Napi::Object context = info[0].As<Napi::Object>();
+    Napi::Object input_serialization_format = context.Get("input_serialization_format").As<Napi::Object>();
+    Napi::Buffer headers_buf_obj = context.Get("records_header_buf").As<Napi::Buffer<uint8_t>>();
+    headers_buf = headers_buf_obj.Data();
+    headers_len = headers_buf_obj.Length();
+    std::string query = context.Get("query").ToString();
+    std::string input_format_str = context.Get("input_format").ToString();
+
+    if (!input_format_str.compare("JSON")) {
+        input_format = InputFormat::json;
+    } else if (!input_format_str.compare("CSV")) {
+        input_format = InputFormat::csv;
+    } else {
+        throw Napi::Error::New(info.Env(), XSTR() << "input_format is neither csv nor json");
+    }
+
+    s3select.parse_query(query.c_str());
+    if (!s3select.get_error_description().empty()) {
+        throw Napi::Error::New(info.Env(), XSTR() << "s3select: parse_query failed " << s3select.get_error_description());
+    }
```

Comment:
What will map all the s3select library errors to the correct protocol error types?
See https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#SelectObjectContentErrorCodeList

Reply:
I will look into mapping the. It'll probably need to hard-coded manually. Also there's no guarantee that s3select in aws error codes will match nicely.

Reply:
In any case I would write the error mapping in js, but here we just want to propagate the information from the library inside the thrown js error object - an error is an object, so you can set more properties on it such as an `err.code` string.

Does the library provide structure error codes? or just a description? if there are no error codes enumerated I would raise an issue to the library asking for the error info in a structured manner.

Reply:
S3select lib seem to go by free hard-coded text rather than enums/something structured.
Git-grep on base_s3select_excption brings a bunch of errors with free text as description and nothing else
(eg https://github.com/ceph/s3select/blob/2e96e6feeaf242f2adde4f050bf6a7f89cfd2c0f/include/s3select.h#L1280).

Reply:
better raise an issue on it and ask for proper error codes that can be mapped for s3 compatibility.

Reply:
if you can open an issue and put the link here for future reference it would be great.